File: 0112-path-sum/0112-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        if not root:
            return False

        def inorder(root, pathSum):
            if not root:
                return None

            if not root.left and not root.right:
                if pathSum + root.val == targetSum:
                    return True

            left = inorder(root.left, pathSum + root.val)

            right = inorder(root.right, pathSum + root.val)

            return left or right

        return inorder(root, 0)

            
        

File: 0112-path-sum/0112-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean hasPathSum(TreeNode root, int targetSum) {
        if (root == null) {
          return false;
        }
        // if leaf node
        if (root.left == null && root.right == null && targetSum == root.val) {
          return true;
        }

        boolean leftPath = hasPathSum(root.left, targetSum - root.val);
        boolean rightPath = hasPathSum(root.right, targetSum - root.val);
        return leftPath || rightPath;
    }
}